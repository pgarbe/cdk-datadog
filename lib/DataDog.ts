export enum DataDogSite {
  eu = "datadoghq.eu",
  us = "datadoghq.com"
}

export interface DataDogCredentials {
  readonly datadogApiKey?: string;
  readonly datadogAppKey?: string;
  readonly datadogSite?: DataDogSite;
}