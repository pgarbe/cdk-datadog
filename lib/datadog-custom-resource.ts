import * as cdk from '@aws-cdk/core';

export interface DataDogAccountIntegrationProps {
}

export class DataDogAccountIntegration extends cdk.Construct {

  constructor(scope: cdk.Construct, id: string) {
    super(scope, id);
  
  //   const supportedResource = [
  //     { 
  //       'TypeName': 'Datadog::Integrations::AWS',
  //       'SchemaHandlerPackage': 's3://datadog-cloudformation-resources/datadog-integrations-aws/datadog-integrations-aws-1.0.1.zip'
  //     },
  //     { 
  //       'TypeName': 'Datadog::Monitors::Monitor',
  //       'SchemaHandlerPackage': 's3://datadog-cloudformation-resources/datadog-monitors-monitor/datadog-monitors-monitor-1.0.1.zip'
  //     },
  //     { 
  //       'TypeName': 'Datadog::Monitors::Downtime',
  //       'SchemaHandlerPackage': 's3://datadog-cloudformation-resources/datadog-monitors-downtime/datadog-monitors-downtime-1.0.1.zip'
  //     },
  //     { 
  //       'TypeName': 'Datadog::IAM::User',
  //       'SchemaHandlerPackage': 's3://datadog-cloudformation-resources/datadog-iam-user/datadog-iam-user-1.0.1.zip'
  //     },
  //   ];
  }

    // TODO: 
    // aws cloudformation register-type \
    //   --type RESOURCE \
    //   --type-name "Datadog::Integrations::AWS" \
    //   --schema-handler-package "s3://datadog-cloudformation-resources/datadog-integrations-aws/datadog-integrations-aws-1.0.1.zip" \
    //   --profile pgarbe

    // Wait for completion and get version from 'TypeVersionArn'
    // aws cloudformation describe-type-registration   --registration-token "3349b1c6-6ef4-47ca-a80b-ec517c995c84" --profile pgarbe

    // Get latest version
    // aws cloudformation list-type-versions \
    //   --type RESOURCE \
    //   --type-name "Datadog::Integrations::AWS" \
    //   --profile pgarbe

    // aws cloudformation set-type-default-version \
    // --region "<REGION>" \
    // --type RESOURCE \
    // --version-id <VERSION_ID> \
    // --type-name "<DATADOG_RESOURCE_NAME>"
}


// aws cloudformation deregister-type \
//   --arn "arn:aws:cloudformation:eu-west-1:424144556073:type/resource/Datadog-Integrations-AWS/00000001" \
//   --profile pgarbe

// aws cloudformation deregister-type \
//   --type "RESOURCE" \
//   --type-name "Datadog::Integrations::AWS" \
//   --version-id "00000003" \
//   --profile pgarbe
