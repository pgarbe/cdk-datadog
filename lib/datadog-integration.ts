import * as cdk from '@aws-cdk/core';
import { Stack } from '@aws-cdk/core';
import { DataDogCredentials } from './DataDog';

export interface DataDogIntegrationProps {
  readonly roleName: string;
  readonly creds: DataDogCredentials;
}

export class DataDogIntegration extends cdk.Construct {

  constructor(scope: cdk.Construct, id: string, props: DataDogIntegrationProps) {
    super(scope, id);

    new cdk.CfnResource(this, 'DataDogAwsIntegration', {
      type: 'Datadog::Integrations::AWS',
      properties: {
        AccountID: Stack.of(this).account,
        RoleName: props.roleName,
        FilterTags: [], // ["filter:thisTag"];
        HostTags: [], // ["env:staging", "account:123456"]
        AccountSpecificNamespaceRules: {}, // {"api_gateway": true, "route53": false},

        DatadogCredentials: {
          ApiURL: 'https://api.' + props.creds.datadogSite,
          ApiKey: props.creds.datadogApiKey,
          ApplicationKey: props.creds.datadogAppKey
        }
      }
    });
  }
}
