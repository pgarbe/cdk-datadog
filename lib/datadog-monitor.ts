import * as cdk from '@aws-cdk/core';
import { DataDogCredentials } from './DataDog';

export enum MonitorType {
  anomaly = "query alert",
  apm = "query alert",
  composite = "composite",
  custom = "service check",
  event = "event alert",
  forecast = "query alert",
  host = "service check",
  integration = "query alert", // or service check
  live = "process	process alert",
  logs = "log alert",
  metric = "metric alert",
  network = "service check",
  outlier = "query alert",
  process = "service check",
  watchdog = "event alert",
}

export interface DataDogMonitorProps {
  readonly creds: DataDogCredentials;
  readonly type: MonitorType;
  readonly query: string;
  readonly name: string;
  readonly message: string;
  // todo: tags
  // todo: readonly options: MonitorOptions;
}

export class DataDogMonitor extends cdk.Construct {

  constructor(scope: cdk.Construct, id: string, props: DataDogMonitorProps) {
    super(scope, id);

    new cdk.CfnResource(this, 'DataDogMonitor', {
      type: 'Datadog::Monitors::Monitor',
      properties: {
        Type: props.type,
        Query: props.query,
        Name: props.name,
        DatadogCredentials: {
          ApiURL: 'https://' + props.creds.datadogSite,
          ApiKey: props.creds.datadogApiKey,
          ApplicationKey: props.creds.datadogAppKey
        }
      }
    });
  }
}

// export class AppStack extends cdk.Stack {
//     constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
//         super(scope, id, props);

//         // ssm.StringParameter.valueForSecureStringParameter(this, '/landingzone/datadog/eu/api-key', 1);
//         var apiKey = new cdk.CfnParameter(this, 'DataDogApiKey', { 
//           default: '/landingzone/datadog/eu/api-key',
//           noEcho: true,
//           type: 'AWS::SSM::Parameter::Value<String>'
//         });
//         var appKey = new cdk.CfnParameter(this, 'DataDogAppKey', { 
//           default: '/landingzone/datadog/eu/application-key',
//           noEcho: true,
//           type: 'AWS::SSM::Parameter::Value<String>'
//         });

//         // var apiKey = ssm.StringParameter.fromStringParameterName(this, 'DataDogApiKey', '/landingzone/datadog/eu/api-key');
//         // var appKey = ssm.StringParameter.fromStringParameterName(this, 'DataDogAppKey', '/landingzone/datadog/eu/app-key');

//         var datadogMonitor = new DataDogMonitor(this, 'MyDataDogMonitor', {
//           datadogApiKey: apiKey.valueAsString,
//           datadogAppKey: appKey.valueAsString,
//           name: 'my monitor',
//           message: 'some message',
//           datadogSite: DataDogSite.eu,
//           query: 'avg(last_1h):sum:system.cpu.system{host:host0} > 100',
//           type: MonitorType.anomaly
//         });

//         // You should stop here
//     }
// }