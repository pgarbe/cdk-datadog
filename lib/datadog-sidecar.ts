import * as cdk from '@aws-cdk/core';
import { DataDogSite } from './DataDog';
import * as ecs from '@aws-cdk/aws-ecs';
import { TaskDefinition } from '@aws-cdk/aws-ecs';

// declare module '@aws-cdk/aws-ecs' {
  // interface TaskDefinition {
  //   withDataDogSidecar(this: TaskDefinition, props: DataDogIntegrationProps): void;
  // }
// }

// TaskDefinition.prototype.withDataDogSidecar = function withDataDogSidecar(this: TaskDefinition, props: DataDogIntegrationProps) {
//   new DatadogSidecar((this as any), 'DataDogAgent', this, props);
// }


export interface DataDogIntegrationProps {
  /**
   * The secret that stores the Datadog API key.
   */
  readonly datadogApiKey: ecs.Secret;

  /**
   * Destination site for your metrics, traces, and logs. Default: DataDogSite.us
   */
  readonly datadogSite?: DataDogSite | undefined;

  /**
   * Enable trace collection with the Trace Agent. Default: true
   */
  readonly apmEnabled?: Boolean | undefined

  /**
   * Enable collection of JMX metrics by Java Tracing Agent. Default: true
   */
  readonly jmxFetchEnabled?: Boolean | undefined

  /**
   * The logging configuration for Datadog container.
   */
  readonly logging?: ecs.LogDriver | undefined;
}

export class DatadogSidecar extends cdk.Construct implements cdk.ITaggable {

  /**
   * Creates a DataDog sidecar to an existing task definition.
   */
  public static addToTaskDefinition(taskDefinition: TaskDefinition, props: DataDogIntegrationProps): cdk.Construct {
    return new DatadogSidecar(taskDefinition.stack, "DataDogSideCar", taskDefinition, props);
  }

  tags: cdk.TagManager;
  private taskDefinition: ecs.TaskDefinition;
  private props: DataDogIntegrationProps

  constructor(scope: cdk.Construct, id: string, taskDefinition: TaskDefinition, props: DataDogIntegrationProps) {
    super(scope, id);
    this.taskDefinition = taskDefinition;
    this.props = props;

    // https://play-with-cdk.com?s=ed63e7f28e4101404cd917241159b8d5
    this.tags = new cdk.TagManager(cdk.TagType.KEY_VALUE, 'myresourceTypeName'); // TODO: Allow custom tags as well (as 3rd parameter)
  }

  /**
   * The resources of this Constructs are rendered here, as environment variables of ContainerDefinition can only be set in constructor,
   * but tags can beapplied later. So this is a safe place to render all needed resources with all .
   */
  prepare() {

    let apmEnabled = (this.props.apmEnabled === undefined) ? true : this.props.apmEnabled;
    let jmxFetchEnabled = (this.props.jmxFetchEnabled === undefined) ? true : this.props.jmxFetchEnabled;
    let datadogSite = (this.props.datadogSite === undefined) ? DataDogSite.us : this.props.datadogSite;
    let datadogTags = this.tags.renderTags().map((tag: any)  => {
      return '"' + tag.Key + ":" + tag.Value + '"'
    });

    // { [key: string]: string; }
    let env = {
      // ...this.tags.renderTags(), // TODO: Format must be DD_TAGS: x.y Validate through tests!
      ECS_FARGATE: 'true',
      DD_APM_ENABLED: apmEnabled.toString(),
      DD_JMXFETCH_ENABLED: jmxFetchEnabled.toString(),
      DD_SITE: datadogSite.toString(),
    };

    if (datadogTags.length > 0) {
      (env as any)['DD_DOGSTATSD_TAGS'] = "[" + datadogTags.toString() + "]"
    }

    const appContainer = this.taskDefinition.defaultContainer;

    const datadog = this.taskDefinition.addContainer('dd-agent', {
      image: ecs.ContainerImage.fromRegistry('datadog/docker-dd-agent'),
      cpu: 64,
      memoryLimitMiB: 128,
      environment: env,
      secrets: {
        DD_API_KEY: this.props.datadogApiKey
      },
      essential: false,
    });

    if (appContainer !== undefined) {
      appContainer.addContainerDependencies(
        {
          container: datadog,
          condition: ecs.ContainerDependencyCondition.SUCCESS
        }
      )
    }

    datadog.addPortMappings({
      containerPort: 8126,
      protocol: ecs.Protocol.TCP,
    });
    datadog.addPortMappings({
      containerPort: 8125,
      protocol: ecs.Protocol.TCP,
    });
  }
}
