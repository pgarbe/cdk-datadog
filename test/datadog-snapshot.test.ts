import { SynthUtils } from '@aws-cdk/assert';
import cdk = require('@aws-cdk/core');
import CdkDatadog = require('../lib/datadog-monitor');
import { SecretValue } from '@aws-cdk/core';
import { DataDogSite } from "../lib/DataDog";

test('Custom Resource created', () => {
  const app = new cdk.App();
  const stack = new cdk.Stack(app, "TestStack");
  const secret = new SecretValue('foo');

  // WHEN
  new CdkDatadog.DataDogMonitor(stack, 'MyTestConstruct', { 
    creds: {
      datadogApiKey: secret.toString(),
      datadogAppKey: secret.toString(),
      datadogSite: DataDogSite.EU,
    },
    name: 'name',
    query: 'query',
    message: 'message',
    type: CdkDatadog.MonitorType.WATCHDOG
  });

  // THEN
  expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
});
