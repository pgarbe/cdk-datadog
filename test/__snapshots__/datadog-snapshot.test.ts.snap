// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Custom Resource created 1`] = `
Object {
  "Resources": Object {
    "MyTestConstructDataDogMonitorF678180F": Object {
      "Properties": Object {
        "DatadogCredentials": Object {
          "ApiKey": "foo",
          "ApiURL": "https://datadoghq.eu",
          "ApplicationKey": "foo",
        },
        "Name": "name",
        "Query": "query",
        "Type": "event alert",
      },
      "Type": "Datadog::Monitors::Monitor",
    },
  },
}
`;
